File: 14-TypeScript/Grey Multi-Tenant/packages/grey-r/R/auth_client.R
#' @title Auth Client
#' @description Authentication client for login, logout, and token refresh
#' @name AuthClient
NULL

#' @importFrom R6 R6Class
#' @export
AuthClient <- R6::R6Class(
  "AuthClient",
  
  public = list(
    #' @description Create a new auth client
    #' @param http_client HttpClient instance
    initialize = function(http_client) {
      private$http <- http_client
    },
    
    #' @description Login with email and password
    #' @param email User email
    #' @param password User password
    #' @return grey_result with tokens or error
    login = function(email, password) {
      # Validate inputs
      if (is.null(email) || nchar(trimws(email)) == 0) {
        return(result_err(error_validation("Email is required")))
      }
      
      if (is.null(password) || nchar(trimws(password)) == 0) {
        return(result_err(error_validation("Password is required")))
      }
      
      result <- private$http$post("/api/v1/auth/login", list(
        email = email,
        password = password
      ))
      
      # Store access token on success
      if (is_ok(result)) {
        data <- result$data
        if (!is.null(data$access_token)) {
          private$http$set_token(data$access_token)
        }
      }
      
      result
    },
    
    #' @description Logout the current user
    #' @return grey_result
    logout = function() {
      result <- private$http$post("/api/v1/auth/logout")
      
      # Clear token on success
      if (is_ok(result)) {
        private$http$clear_token()
      }
      
      result
    },
    
    #' @description Refresh the access token
    #' @param refresh_token The refresh token
    #' @return grey_result with new tokens or error
    refresh = function(refresh_token) {
      if (is.null(refresh_token) || nchar(trimws(refresh_token)) == 0) {
        return(result_err(error_validation("Refresh token is required")))
      }
      
      result <- private$http$post("/api/v1/auth/refresh", list(
        refresh_token = refresh_token
      ))
      
      # Update access token on success
      if (is_ok(result)) {
        data <- result$data
        if (!is.null(data$access_token)) {
          private$http$set_token(data$access_token)
        }
      }
      
      result
    },
    
    #' @description Manually set the access token
    #' @param token Access token string
    set_token = function(token) {
      private$http$set_token(token)
    }
  ),
  
  private = list(
    http = NULL
  )
)

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-vb/Domain/AuthClient.vb
' Grey SDK for Visual Basic - Auth Client
' Authentication operations: login, logout, refresh.

Imports Grey.Errors
Imports Grey.Http

Namespace Grey.Domain

    ''' <summary>
    ''' Authentication tokens returned from login/refresh.
    ''' </summary>
    Public Class AuthTokens
        ''' <summary>Gets or sets the access token.</summary>
        Public Property AccessToken As String

        ''' <summary>Gets or sets the refresh token.</summary>
        Public Property RefreshToken As String

        ''' <summary>Gets or sets the token expiration in seconds.</summary>
        Public Property ExpiresIn As Integer
    End Class

    ''' <summary>
    ''' Client for authentication operations.
    ''' </summary>
    Public Class AuthClient

        Private ReadOnly _httpClient As GreyHttpClient

        ''' <summary>
        ''' Creates a new AuthClient instance.
        ''' </summary>
        ''' <param name="httpClient">The HTTP client to use.</param>
        Public Sub New(httpClient As GreyHttpClient)
            _httpClient = httpClient
        End Sub

        ''' <summary>
        ''' Authenticates a user with username and password.
        ''' </summary>
        ''' <param name="username">The username.</param>
        ''' <param name="password">The password.</param>
        ''' <returns>Result containing auth tokens on success.</returns>
        Public Async Function LoginAsync(username As String, password As String) As Task(Of Result(Of AuthTokens))
            ' Validate inputs
            If String.IsNullOrWhiteSpace(username) Then
                Return Result(Of AuthTokens).Fail(
                    GreyError.ValidationError("Username is required"))
            End If

            If String.IsNullOrWhiteSpace(password) Then
                Return Result(Of AuthTokens).Fail(
                    GreyError.ValidationError("Password is required"))
            End If

            Dim body = New With {
                .username = username,
                .password = password
            }

            Return Await _httpClient.PostAsync(Of AuthTokens)("/api/auth/login", body)
        End Function

        ''' <summary>
        ''' Logs out the current user.
        ''' </summary>
        ''' <returns>Result indicating success or failure.</returns>
        Public Async Function LogoutAsync() As Task(Of Result(Of Boolean))
            Return Await _httpClient.PostAsync(Of Boolean)("/api/auth/logout", Nothing)
        End Function

        ''' <summary>
        ''' Refreshes the authentication token.
        ''' </summary>
        ''' <param name="refreshToken">The refresh token.</param>
        ''' <returns>Result containing new auth tokens on success.</returns>
        Public Async Function RefreshAsync(refreshToken As String) As Task(Of Result(Of AuthTokens))
            ' Validate inputs
            If String.IsNullOrWhiteSpace(refreshToken) Then
                Return Result(Of AuthTokens).Fail(
                    GreyError.ValidationError("Refresh token is required"))
            End If

            Dim body = New With {
                .refreshToken = refreshToken
            }

            Return Await _httpClient.PostAsync(Of AuthTokens)("/api/auth/refresh", body)
        End Function

    End Class

End Namespace
